crypto/bls12/bls12.go

```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+	defer func() {
+		if !valid {
+			bls.mods.Logger().Debugf("Invalid proof-of-possession for replica %d", replica.ID())
+		}
+	}()
+
+	b, ok := replica.Metadata()[popMetadataKey]
 	if !ok {
+		bls.mods.Logger().Debugf("Missing proof-of-possession for replica: %d", replica.ID())
 		return false
 	}
-	pubKeys := make([]*PublicKey, 0)
-	sig.participants.ForEach(func(id hotstuff.ID) {
-		replica, ok := bc.mods.Configuration().Replica(id)
-		if !ok {
-			return
-		}
-		pubKeys = append(pubKeys, replica.PublicKey().(*PublicKey))
-	})
-	ps, err := bls12.NewG2().HashToCurve(hash[:], domain)
+
+	var key strings.Builder
+	key.WriteString(b)
+	_, _ = key.Write(replica.PublicKey().(*PublicKey).ToBytes())
+
+	bls.mut.RLock()
+	valid, ok = bls.popCache[key.String()]
+	bls.mut.RUnlock()
+	if ok {
+		return valid
+	}
+
+	p, err := bls12.NewG2().FromCompressed([]byte(b))
 	if err != nil {
-		bc.mods.Logger().Error(err)
 		return false
 	}
-	if len(pubKeys) < bc.mods.Configuration().QuorumSize() {
+
+	valid = bls.popVerify(replica.PublicKey().(*PublicKey), p)
+
+	bls.mut.Lock()
+	bls.popCache[key.String()] = valid
+	bls.mut.Unlock()
+
+	return valid
+}
+
+func (bls *bls12Base) coreAggregateVerify(publicKeys []*PublicKey, messages [][]byte, signature *bls12.PointG2) bool {
+	n := len(publicKeys)
+	if n != len(messages) && n <= 1 {
+		return false
+	}
+
+	if !bls.subgroupCheck(signature) {
 		return false
 	}
+
 	engine := bls12.NewEngine()
-	engine.AddPairInv(&bls12.G1One, &sig.sig)
-	for _, pub := range pubKeys {
-		engine.AddPair(pub.p, ps)
+
+	for i := 0; i < n; i++ {
+		q, err := engine.G2.HashToCurve(messages[i], domain)
+		if err != nil {
+			return false
+		}
+		engine.AddPair(publicKeys[i].p, q)
 	}
+
+	engine.AddPairInv(&bls12.G1One, signature)
 	return engine.Result().IsOne()
 }
 
-// VerifyThresholdSignatureForMessageSet verifies a threshold signature against a set of message hashes.
-func (bc *bls12Crypto) VerifyThresholdSignatureForMessageSet(signature consensus.ThresholdSignature, hashes map[hotstuff.ID]consensus.Hash) bool {
+func (bls *bls12Base) aggregateVerify(publicKeys []*PublicKey, messages [][]byte, signature *bls12.PointG2) bool {
+	set := make(map[string]struct{})
+	for _, m := range messages {
+		set[string(m)] = struct{}{}
+	}
+	return len(messages) == len(set) && bls.coreAggregateVerify(publicKeys, messages, signature)
+}
+
+func (bls *bls12Base) fastAggregateVerify(publicKeys []*PublicKey, message []byte, signature *bls12.PointG2) bool {
+	engine := bls12.NewEngine()
+	var aggregate bls12.PointG1
+	for _, pk := range publicKeys {
+		engine.G1.Add(&aggregate, &aggregate, pk.p)
+	}
+	return bls.coreVerify(&PublicKey{p: &aggregate}, message, signature, domain)
+}
+
+// Sign creates a cryptographic signature of the given messsage.
+func (bls *bls12Base) Sign(message []byte) (signature consensus.QuorumSignature, err error) {
+	// TODO: consider passing message to this function instead of hash
```

Comment:
Still valid TODO? The input `message` seems to indicate it is already a message.